Google.Cloud.AspNetCore/Google.Cloud.AspNetCore.Firestore.DistributedCache/FirestoreCache.cs

```diff
@@ -0,0 +1,323 @@
+            CacheDoc doc = snapshot.ConvertTo<CacheDoc>();
+            var now = DateTime.UtcNow;
+            if (doc.AbsoluteExpiration.HasValue &&
+                doc.AbsoluteExpiration.Value < now)
+            {
+                return null;
+            }
+            var slidingExpiration = doc.LastRefresh.GetValueOrDefault()
+                + TimeSpan.FromSeconds(doc.SlidingExpirationSeconds.GetValueOrDefault());
+            if (slidingExpiration < now)
+            {
+                return null;
+            }
+            return doc.Value;
+        }
+        void IDistributedCache.Refresh(string key)
+        {
+            try
+            {
+                _cacheEntries.Document(key)
+                    .UpdateAsync("LastRefresh", DateTime.UtcNow).Wait();
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        async Task IDistributedCache.RefreshAsync(string key, CancellationToken token)
+        {
+            try
+            {
+                await _cacheEntries.Document(key).UpdateAsync(
+                    "LastRefresh", DateTime.UtcNow, cancellationToken: token);
+            }
+            catch (Grpc.Core.RpcException e)
+            when (e.StatusCode == Grpc.Core.StatusCode.NotFound)
+            {
+                // Curiously, ASP.NET middleware will call Refresh for
+                // cache entries that have never been Set.  That's ok,
+                // but there's nothing for us to do.
+            }
+        }
+
+        void IDistributedCache.Remove(string key) =>
+            _cacheEntries.Document(key).DeleteAsync().Wait();
+
+        Task IDistributedCache.RemoveAsync(string key, CancellationToken token) =>
+            _cacheEntries.Document(key).DeleteAsync(cancellationToken: token);
+
+        /// <summary>
+        /// Creates a firestore document that stores the cache value and its
+        /// expiration info, but does not store it in firestore.
+        /// </summary>
+        /// <param name="value">The value to be stored in the cache.</param>
+        /// <param name="options">Expiration info.</param>
+        /// <returns></returns>
+        CacheDoc MakeCacheDoc(byte[] value, DistributedCacheEntryOptions options)
+        {
+            CacheDoc doc = new CacheDoc()
+            {
+                LastRefresh = DateTime.UtcNow,
+                Value = value,
+            };
+            if (options.SlidingExpiration.HasValue)
+            {
+                doc.SlidingExpirationSeconds =
+                    options.SlidingExpiration.Value.TotalSeconds;
+            }
+            else
+            {
+                var forever = DateTime.MaxValue - doc.LastRefresh.Value;

```

Comment:
I don't understand what this is trying to do, at least not for sure. My guess is that you're halving it to avoid overflow when you add it in `ValueFromSnapshot`. That could be avoided by doing subtraction of last refresh from now instead, to find out "time since refresh" instead of "predicted refresh". Alternatively, I'd be tempted to set it to some constant of "1000 years" as "close enough to forever to make no odds, and doesn't require any computation". That would be easier to test.

Reply:
Yes, it was to avoid overflow.  I changed it to 1000 years.